quickwit/quickwit-serve/src/index_api/rest_handler.rs

```diff
@@ -141,6 +146,129 @@ fn get_indexes_metadatas_handler(
         .map(make_response)
 }
 
+
+
+// This structure represents describe Index
+
+#[derive(Serialize, Deserialize)]
+pub struct IndexStats {
+    pub index_id: String,
+    pub index_uri: Uri,
+    pub num_published_splits: usize,
+    pub num_published_docs: u64,
+    pub size_published_docs: Byte,
+    pub timestamp_field_name: Option<String>,
+    pub timestamp_range: Option<(i64, i64)>,
+}
+
+#[utoipa::path(
+    get,
+    tag = "Indexes",
+    path = "/indexes/{index_id}/describe",
+    responses(
+        (status = 200, description = "Successfully fetched stats about Index.", body = IndexStats)
+    ),
+    params(
+        ("index_id" = String, Path, description = "The index ID to retrieve index metadata and split data."),
+    )
+)]
+
+// Get describe index
+async fn describe_index(
+    index_id: String,
+    metastore: Arc<dyn Metastore>,
+) -> Result<IndexStats, MetastoreError> {
+    
+    //get index_metadata struct
+    let index_metadata = metastore.index_metadata(&index_id).await;
+    let index_metadata = match index_metadata{
+        Ok(index_metadata) => index_metadata,
+        Err(e) => return Err(e)
+    };
+
+    //get splits_metadata struct
+    let query = ListSplitsQuery::for_index(&index_id);
+    let splits = metastore.list_splits(query).await;
+    let splits = match splits{
+        Ok(splits) => splits,
+        Err(e) => return Err(e)
+    };
+
+    let published_splits: Vec<Split> = splits
+            .into_iter()
+            .filter(|split| split.split_state == SplitState::Published)
+            .collect();
+    
+    let splits_num_docs = published_splits
+        .iter()
+        .map(|split| split.split_metadata.num_docs as u64)
+        .sorted()
+        .collect_vec();
+
+    let total_num_docs = splits_num_docs.iter().sum::<u64>();
+
+    let splits_bytes = published_splits
+        .iter()
+        .filter(|split| split.split_state == SplitState::Published)
+        .map(|split| split.split_metadata.footer_offsets.end)
+        .sorted()
+        .collect_vec();
+    let total_bytes = splits_bytes.iter().sum::<u64>();
+
+    let timestamp_range = if index_metadata
+        .index_config()
+        .doc_mapping
+        .timestamp_field
+        .is_some()
+    {
+        let time_min = published_splits
+            .iter()
+            .flat_map(|split| split.split_metadata.time_range.clone())
+            .map(|time_range| *time_range.start())
+            .min();
+        let time_max = published_splits
+            .iter()
+            .flat_map(|split| split.split_metadata.time_range.clone())
+            .map(|time_range| *time_range.end())
+            .max();
+        if let (Some(time_min), Some(time_max)) = (time_min, time_max) {
+            Some((time_min, time_max))
+        } else {
+            None
+        }
+    } else {
+        None
+    };
```

Comment:
I think you may find a simple for loop here will simplify a lot of these iterations.

Reply:
what did you mean @ChillFish8?

Reply:
We have a set of iterator chains here which all essentially iterate over the same set of values several times.

I think it would be simpler to make it a for loop iterating over the published splits in order to produce the total doc count, total byte count, etc...

Reply:
yes indeed